### PR TITLE
server/auth: implement user & user manager.

### PR DESCRIPTION
**Pull request by PhotonQuantum**
Signed-off-by: PhotonQuantum <self@lightquantum.me>

### What problem does this PR solve?

ref: https://github.com/tikv/tikv/issues/8621
server/auth: implement user & user manager.

Related PR: #3256 

### What is changed and how it works?

server/auth `userManager` is implemented to handle user-related CRUD operations, including querying users in memory, loading them from etcd and persisting them to etcd.

### Check List

Tests

- Unit test

Code changes

- Has persistent data change (etcd)

### Release note

- No release note

### CODE REVIEW

File: server/auth/auth.go [hunk 1]
```diff
@@ -24,18 +24,270 @@ import (
...
+	user, err := NewUser(name, GenerateHash(password))
+	if err != nil {
+		return err
+	}
+
+	userJSON, err := json.Marshal(user)
+	if err != nil {
+		return errs.ErrJSONMarshal.Wrap(err).GenWithStackByCause()
+	}
+	userPath := path.Join(userPrefix, name)
```
Comment (JmPotato) [inline server/auth/auth.go:99]: I saw the use of this function for multi-times, maybe we can make it become a common method?

File: server/auth/auth.go [hunk 2]
```diff
@@ -24,18 +24,270 @@ import (
...
+}
+
 type roleManager struct {
 	kv    kv.Base
 	mu    sync.RWMutex
 	roles map[string]*Role
 }
 
-// newRoleManager creates a new roleManager.
+// newRoleManager creates a new roleManager for debug purposes.
```
Comment (Yisaer) [inline server/auth/auth.go:290]:
```suggestion
// newRoleManager creates a new roleManager.
```

File: server/auth/auth.go [hunk 3]
```diff
@@ -24,18 +24,270 @@ import (
...
+
+	if ok := user.appendRole(role); !ok {
+		return errs.ErrUserHasRole.FastGenByArgs(name, role)
+	}
+
+	userJSON, err := json.Marshal(role)
+	if err != nil {
+		return errs.ErrJSONMarshal.Wrap(err).GenWithStackByCause()
+	}
+	userPath := path.Join(userPrefix, name)
```
Comment (Yisaer) [inline server/auth/auth.go:214]: I think we can define a function in util like `GetUserPath` and `GetRolePath`.

File: server/auth/auth.go [hunk 4]
```diff
@@ -24,18 +24,270 @@ import (
...
+
+	return nil
+}
+
+// UpdateCache refreshes in-memory cache of users.
+func (m *userManager) UpdateCache() error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	userPath := strings.Join([]string{userPrefix, ""}, "/")
```
Comment (Yisaer) [inline server/auth/auth.go:265]: I see different methods to build userPath

**Review by Yisaer — COMMENTED** [review]
plz address the comment

File: server/auth/auth.go [hunk 5]
```diff
@@ -24,18 +24,270 @@ import (
...
+	user.Hash = GenerateHash(password)
+
+	// Update user in kv
+	userJSON, err := json.Marshal(user)
+	if err != nil {
+		return errs.ErrJSONMarshal.Wrap(err).GenWithStackByCause()
+	}
+	userPath := path.Join(userPrefix, name)
+
+	err = m.kv.Save(userPath, string(userJSON))
```
Comment (rleungx) [inline server/auth/auth.go:159]: What if we have something wrong when saving kv, will the password be rollback?

**Comment by gregwebs** [comment]
All the auth information is cached in memory. What happens when the PD leader changes or when a different PD is contacted? I am wondering if we can just rely on etcd to cache this information as it sees fit?

I am also wondering if we could be delegating this RBAC to the RBAC that is already [available in etcd](https://etcd.io/docs/v3.3.12/op-guide/authentication/).

File: server/auth/auth.go (hunk 3 above)
Comment (PhotonQuantum) [reply server/auth/auth.go]: Extracted to util.

File: server/auth/auth.go (hunk 4 above)
Comment (PhotonQuantum) [reply server/auth/auth.go]: Using this method in `GetUserPath` because it can handle trailing slash correctly.

File: server/auth/auth.go (hunk 5 above)
Comment (PhotonQuantum) [reply server/auth/auth.go:154]: Wow, it's a bug! Thank you for pointing it out!

File: server/auth/user.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+	ok := validateName(username)
+	if !ok {
+		return nil, errs.ErrInvalidUserName.FastGenByArgs(username)
+	}
+
+	return &User{Username: username, Hash: hash, RoleKeys: make([]string, 0)}, nil
+}
+
+// NewUserFromJSON safely deserialize a json string to a user instance.
+func NewUserFromJSON(j string) (*User, error) {
```
[the remark this answers is not in the document]
Comment (PhotonQuantum) [reply server/auth/user.go]: Renamed.

File: server/auth/user.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+	"encoding/json"
+
+	"github.com/tikv/pd/pkg/errs"
+)
+
+// User records user info.
+// Read-Only once created.
+type User struct {
+	Username string   `json:"username"`
+	Hash     string   `json:"hash"`
```
[the remark this answers is not in the document]
Comment (PhotonQuantum) [reply server/auth/user.go]: Added.

File: server/auth/auth.go (hunk 1 above)
Comment (PhotonQuantum) [reply server/auth/auth.go]: Extracted into `GetUserPath` and `GetRolePath`.

File: server/auth/auth.go (hunk 2 above)
Comment (PhotonQuantum) [reply server/auth/auth.go]: Removed.

**Comment by gregwebs** [comment]
Any response to my comments? Does the memory caching work when more than one PD is used?

Since this PR is just to a branch I think we can merge it once tests are passing.

**Comment by PhotonQuantum** [comment]
> Any response to my comments?

Sorry for the late reply! I'm quite busy with my studies these days, so I may not respond in time 😢. But I will try to do so as long as I can.

> All the auth information is cached in memory. What happens when the PD leader changes or when a different PD is contacted?

I believe we may refresh the cache when a new PD leader is elected, but this won't work when a follower is contacted. I'm thinking about redirecting all requests to the leader when auth is enabled, for now.

> I am wondering if we can just rely on etcd to cache this information as it sees fit.

I don't like it because this will reduce the reflexivity of the auth system, and slow down the authentication process (accessing etcd per request is obviously slower than reading from in4-memory cache.).

@Yisaer What's your opinion on this?

**Comment by Yisaer** [comment]
> What happens when the PD leader changes or when a different PD is contacted?

I think we can load the data from etcd while the member is elected as leader. Also we could create a `gouroutine` to refresh the data from etcd to memory to keep consistency.

> delegating this RBAC to the RBAC that is already available in etcd.

delegating RBAC to etcd RBAC strategy means all the resources the PD holding should saved into etcd while etcd is not the only stored backed (pd also use sqlite).

**Review by Yisaer — APPROVED** [review]
LGTM

**Comment by gregwebs** [comment]
> Also we could create a gouroutine to refresh the data from etcd to memory to keep consistency.

This will always create inconsistency windows unless we can know if things are out of date. Refresh on leader election and only sending requests to the leader would keep things 100% consistent though.

> I don't like it because this will reduce the reflexivity of the auth system, and slow down the authentication process (accessing etcd per request is obviously slower than reading from in-memory cache.).

We can benchmark perform implementing performance optimizations. Etcd aggressively caches data in memory so I think we may be duplicating this work. Accessing etcd does not necessarily have much overhead since we are using it in an embedded manner.

**Pull request merged**